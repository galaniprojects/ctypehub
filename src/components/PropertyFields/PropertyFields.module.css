.label {
  composes: label from '../Input/Input.module.css';
}

.input {
  composes: component from '../Input/Input.module.css';
}

@media screen and (width >= 30rem) {
  .group {
    display: flex;
    flex-direction: row;
    justify-content: stretch;
    gap: 1rem;
  }

  .group > * {
    flex-grow: 1;
    flex-shrink: 1;
    margin-block-start: 0;
  }

  .group + * {
    margin-block-start: 0;
  }
}

.fieldset {
  composes: fieldset from '../Input/Input.module.css';
}

.array {
  composes: label;
  margin: 0;
  flex-direction: row;
  align-items: center;
}
