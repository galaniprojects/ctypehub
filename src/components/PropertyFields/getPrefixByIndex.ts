export function getPrefixByIndex(index: number) {
  return `property[${index}].`;
}
