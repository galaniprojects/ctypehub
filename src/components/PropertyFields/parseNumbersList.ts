export function parseNumbersList(
  list: string | undefined,
  parse: (input: string) => number,
) {
  if (!list) {
    return undefined;
  }
  const numbersList = list.split(',').map((value) => parse(value));
  if (numbersList.some(Number.isNaN)) {
    const error = `Cannot parse as list of numbers: ${list}`;
    alert(error);
    throw new Error(error);
  }
  return numbersList;
}
