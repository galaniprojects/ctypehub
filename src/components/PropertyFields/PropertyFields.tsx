import type { PropertyType } from './getProperties';

import { ChangeEvent, Fragment, useCallback, useState } from 'react';

import styles from './PropertyFields.module.css';

import { getPrefixByIndex } from './getPrefixByIndex';

export function PropertyFields({ index }: { index: number }) {
  const prefix = getPrefixByIndex(index);

  const [isArray, setIsArray] = useState(false);
  const handleArrayChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) =>
      setIsArray(event.currentTarget.checked),
    [],
  );

  const [isEnum, setIsEnum] = useState<boolean>();
  const handleIsEnumChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) =>
      setIsEnum(event.currentTarget.value === 'enum'),
    [],
  );

  const [type, setType] = useState<PropertyType>('string');
  const handleTypeChange = useCallback(
    (event: ChangeEvent<HTMLSelectElement>) => {
      const newType = event.currentTarget.value as PropertyType;
      setType(newType);
      if (!['string', 'integer', 'number'].includes(newType)) {
        setIsEnum(undefined);
      }
    },
    [],
  );

  return (
    <Fragment>
      <p>
        <label className={styles.label}>
          Name:
          <input name={`${prefix}name`} className={styles.input} required />
        </label>
      </p>
      <p>
        <label className={styles.label}>
          Type:
          <select
            name={`${prefix}type`}
            className={styles.input}
            onChange={handleTypeChange}
          >
            <option value="string">string</option>
            <option value="integer">integer</option>
            <option value="number">number</option>
            <option value="boolean">boolean</option>
            <option value="reference">reference</option>
          </select>
        </label>
      </p>

      {type === 'reference' && (
        <p>
          <label className={styles.label}>
            $id of Referenced Claim Type or one of its properties
            <input
              name={`${prefix}$ref`}
              className={styles.input}
              placeholder="kilt:ctype:0xâ€¦"
              pattern="kilt:ctype:0x[0-9a-f]+(#/properties/.+)?"
              required
            />
          </label>
        </p>
      )}

      {['string', 'integer', 'number'].includes(type) && (
        <fieldset className={styles.fieldset}>
          <legend>Limitations (optional)</legend>

          <p className={styles.group}>
            <label className={styles.array}>
              <input
                name={`${prefix}limitations`}
                type="radio"
                value="any"
                onChange={handleIsEnumChange}
              />
              any matching value
            </label>
            <label className={styles.array}>
              <input
                name={`${prefix}limitations`}
                type="radio"
                value="enum"
                onChange={handleIsEnumChange}
              />
              fixed set of values
            </label>
          </p>

          {isEnum === false && type === 'string' && (
            <Fragment>
              <div className={styles.group}>
                <p>
                  <label className={styles.label}>
                    Minimum length (optional):
                    <input
                      name={`${prefix}minLength`}
                      type="number"
                      min={0}
                      step={1}
                      className={styles.input}
                    />
                  </label>
                </p>
                <p>
                  <label className={styles.label}>
                    Maximum length (optional):
                    <input
                      name={`${prefix}maxLength`}
                      type="number"
                      min={0}
                      step={1}
                      className={styles.input}
                    />
                  </label>
                </p>
              </div>
              <p>
                <label className={styles.label}>
                  Format (optional):
                  <select name={`${prefix}format`} className={styles.input}>
                    <option value="">unset</option>
                    <option value="date">date</option>
                    <option value="time">time</option>
                    <option value="uri">URI</option>
                  </select>
                </label>
              </p>
            </Fragment>
          )}

          {isEnum === false && ['integer', 'number'].includes(type) && (
            <div className={styles.group}>
              <p>
                <label className={styles.label}>
                  Minimum value (optional):
                  <input
                    name={`${prefix}minimum`}
                    type="number"
                    className={styles.input}
                    step="any"
                  />
                </label>
              </p>
              <p>
                <label className={styles.label}>
                  Maximum value (optional):
                  <input
                    name={`${prefix}maximum`}
                    type="number"
                    className={styles.input}
                    step="any"
                  />
                </label>
              </p>
            </div>
          )}

          {isEnum && (
            <p>
              <label className={styles.label}>
                Comma-separated list of allowed values, no spaces (optional):
                <input
                  name={`${prefix}enum`}
                  className={styles.input}
                  pattern="[^,](,.+)*"
                />
              </label>
            </p>
          )}
        </fieldset>
      )}

      <p>
        <label className={styles.array}>
          <input
            name={`${prefix}array`}
            type="checkbox"
            value="array"
            onChange={handleArrayChange}
          />
          is an array of such values
        </label>
      </p>

      {isArray && (
        <div className={styles.group}>
          <p>
            <label className={styles.label}>
              Minimum number of values (optional):
              <input
                name={`${prefix}minItems`}
                type="number"
                min={0}
                step={1}
                className={styles.input}
              />
            </label>
          </p>
          <p>
            <label className={styles.label}>
              Maximum number of values (optional):
              <input
                name={`${prefix}maxItems`}
                type="number"
                min={0}
                step={1}
                className={styles.input}
              />
            </label>
          </p>
        </div>
      )}
    </Fragment>
  );
}
