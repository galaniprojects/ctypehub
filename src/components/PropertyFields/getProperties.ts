import type { ICType } from '@kiltprotocol/sdk-js';

import { offsets } from '../../utilities/offsets';

import { getPrefixByIndex } from './getPrefixByIndex';
import { parseNumbersList } from './parseNumbersList';

export type PropertyType =
  | 'string'
  | 'integer'
  | 'number'
  | 'boolean'
  | 'reference';

export function getProperties(
  count: number,
  allValues: Array<[string, string]>,
): ICType['properties'] {
  const rawProperties = offsets(count).map((index) => {
    const prefix = getPrefixByIndex(index);
    const matching = allValues.filter(
      ([name, value]) => name.startsWith(prefix) && value !== '',
    );
    return Object.fromEntries(
      matching.map(([name, value]) => [name.replace(prefix, ''), value]),
    );
  });

  // I couldnâ€™t fix that
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return Object.fromEntries(
    rawProperties.map((property) => {
      const type = property.type as PropertyType;

      let data;
      if (type === 'reference') {
        data = { $ref: property.$ref };
      }

      if (type === 'boolean') {
        data = { type };
      }

      if (type === 'string') {
        const { format, minLength, maxLength, enum: list } = property;
        data = {
          type,
          ...(format && { format }),
          ...(minLength && { minLength: parseInt(minLength) }),
          ...(maxLength && { maxLength: parseInt(maxLength) }),
          ...(list && { enum: list.split(',') }),
        };
      }

      if (['integer', 'number'].includes(type)) {
        const { minimum, maximum, enum: list } = property;

        const parse = type === 'integer' ? parseInt : parseFloat;
        const numbersList = parseNumbersList(list, parse);

        data = {
          type,
          ...(minimum && { minimum: parseFloat(minimum) }),
          ...(maximum && { maximum: parseFloat(maximum) }),
          ...(numbersList && { enum: numbersList }),
        };
      }

      const { name, array, minItems, maxItems } = property;
      if (!array) {
        return [name, data];
      }

      return [
        name,
        {
          type: 'array',
          items: data,
          ...(minItems && { minItems: parseInt(minItems) }),
          ...(maxItems && { maxItems: parseInt(maxItems) }),
        },
      ];
    }),
  );
}
