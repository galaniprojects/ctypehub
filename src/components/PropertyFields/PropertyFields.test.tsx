// @vitest-environment jsdom

import { describe, expect, it } from 'vitest';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { PropertyFields } from './PropertyFields';

describe('<PropertyFields />', () => {
  it('renders correctly', async () => {
    const { container, getByLabelText } = render(<PropertyFields index={0} />);
    expect(container).toMatchSnapshot('string');

    await userEvent.click(getByLabelText('any matching value'));
    expect(container).toMatchSnapshot('string limitations');

    const name = getByLabelText('Name:');
    await userEvent.type(name, 'Name');

    const select = getByLabelText('Type:');

    await userEvent.selectOptions(select, 'boolean');
    expect(container).toMatchSnapshot('boolean');

    await userEvent.selectOptions(select, 'reference');
    expect(container).toMatchSnapshot('reference');

    await userEvent.selectOptions(select, 'number');
    expect(container).toMatchSnapshot('number');

    await userEvent.selectOptions(select, 'integer');
    expect(container).toMatchSnapshot('integer');

    await userEvent.click(getByLabelText('any matching value'));
    expect(container).toMatchSnapshot('integer limitations');

    await userEvent.click(getByLabelText('fixed set of values'));
    expect(container).toMatchSnapshot('integer enum');

    const array = getByLabelText(/array/i);
    await userEvent.click(array);
    expect(container).toMatchSnapshot('integer array');
  });
});
