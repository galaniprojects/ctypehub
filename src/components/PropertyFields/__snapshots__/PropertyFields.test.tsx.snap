// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PropertyFields /> > renders correctly > boolean 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > integer 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > integer array 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
    <p>
      <label
        class="_label_8bc908"
      >
        Comma-separated list of allowed values, no spaces (optional):
        <input
          class="_input_8bc908"
          name="property[0].enum"
          pattern="[^,](,.+)*"
        />
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
  <div
    class="_group_8bc908"
  >
    <p>
      <label
        class="_label_8bc908"
      >
        Minimum number of values (optional):
        <input
          class="_input_8bc908"
          min="0"
          name="property[0].minItems"
          step="1"
          type="number"
        />
      </label>
    </p>
    <p>
      <label
        class="_label_8bc908"
      >
        Maximum number of values (optional):
        <input
          class="_input_8bc908"
          min="0"
          name="property[0].maxItems"
          step="1"
          type="number"
        />
      </label>
    </p>
  </div>
</div>
`;

exports[`<PropertyFields /> > renders correctly > integer enum 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
    <p>
      <label
        class="_label_8bc908"
      >
        Comma-separated list of allowed values, no spaces (optional):
        <input
          class="_input_8bc908"
          name="property[0].enum"
          pattern="[^,](,.+)*"
        />
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > integer limitations 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
    <div
      class="_group_8bc908"
    >
      <p>
        <label
          class="_label_8bc908"
        >
          Minimum value (optional):
          <input
            class="_input_8bc908"
            name="property[0].minimum"
            step="any"
            type="number"
          />
        </label>
      </p>
      <p>
        <label
          class="_label_8bc908"
        >
          Maximum value (optional):
          <input
            class="_input_8bc908"
            name="property[0].maximum"
            step="any"
            type="number"
          />
        </label>
      </p>
    </div>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > number 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > reference 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      $id of Referenced Claim Type or one of its properties
      <input
        class="_input_8bc908"
        name="property[0].$ref"
        pattern="kilt:ctype:0x[0-9a-f]+(#/properties/.+)?"
        placeholder="kilt:ctype:0x…"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > string 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;

exports[`<PropertyFields /> > renders correctly > string limitations 1`] = `
<div>
  <p>
    <label
      class="_label_8bc908"
    >
      Name:
      <input
        class="_input_8bc908"
        name="property[0].name"
        required=""
      />
    </label>
  </p>
  <p>
    <label
      class="_label_8bc908"
    >
      Type:
      <select
        class="_input_8bc908"
        name="property[0].type"
      >
        <option
          value="string"
        >
          string
        </option>
        <option
          value="integer"
        >
          integer
        </option>
        <option
          value="number"
        >
          number
        </option>
        <option
          value="boolean"
        >
          boolean
        </option>
        <option
          value="reference"
        >
          reference
        </option>
      </select>
    </label>
  </p>
  <fieldset
    class="_fieldset_8bc908"
  >
    <legend>
      Limitations (optional)
    </legend>
    <p
      class="_group_8bc908"
    >
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="any"
        />
        any matching value
      </label>
      <label
        class="_array_8bc908"
      >
        <input
          name="property[0].limitations"
          type="radio"
          value="enum"
        />
        fixed set of values
      </label>
    </p>
    <div
      class="_group_8bc908"
    >
      <p>
        <label
          class="_label_8bc908"
        >
          Minimum length (optional):
          <input
            class="_input_8bc908"
            min="0"
            name="property[0].minLength"
            step="1"
            type="number"
          />
        </label>
      </p>
      <p>
        <label
          class="_label_8bc908"
        >
          Maximum length (optional):
          <input
            class="_input_8bc908"
            min="0"
            name="property[0].maxLength"
            step="1"
            type="number"
          />
        </label>
      </p>
    </div>
    <p>
      <label
        class="_label_8bc908"
      >
        Format (optional):
        <select
          class="_input_8bc908"
          name="property[0].format"
        >
          <option
            value=""
          >
            unset
          </option>
          <option
            value="date"
          >
            date
          </option>
          <option
            value="time"
          >
            time
          </option>
          <option
            value="uri"
          >
            URI
          </option>
        </select>
      </label>
    </p>
  </fieldset>
  <p>
    <label
      class="_array_8bc908"
    >
      <input
        name="property[0].array"
        type="checkbox"
        value="array"
      />
      is an array of such values
    </label>
  </p>
</div>
`;
